File: tests/abi/ibstream_layout.cpp
#include "pluginterfaces/base/ibstream.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(TYPE) \
	std::printf (#TYPE " size %zu align %zu\n", sizeof (TYPE), alignof (TYPE))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	PRINT_TYPE (Steinberg::IBStream);
	PRINT_TYPE (Steinberg::ISizeableStream);
	std::printf ("IBStream.kIBSeekSet %d\n", Steinberg::IBStream::kIBSeekSet);
	std::printf ("IBStream.kIBSeekCur %d\n", Steinberg::IBStream::kIBSeekCur);
	std::printf ("IBStream.kIBSeekEnd %d\n", Steinberg::IBStream::kIBSeekEnd);
	print_iid ("IBStream", Steinberg::IBStream_iid);
	print_iid ("ISizeableStream", Steinberg::ISizeableStream_iid);
	return 0;
}

File: tools/ibstream_layout.zig
const std = @import("std");
const ibstream = @import("zig-vst3").pluginterfaces.base.ibstream;

pub fn main(init: std.process.Init) !void {
    var stdout_buffer: [4096]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(init.io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    defer stdout.flush() catch {};
    try printType(stdout, "Steinberg::IBStream", ibstream.IBStream);
    try printType(stdout, "Steinberg::ISizeableStream", ibstream.ISizeableStream);
    try stdout.print("IBStream.kIBSeekSet {}\n", .{@intFromEnum(ibstream.IStreamSeekMode.kIBSeekSet)});
    try stdout.print("IBStream.kIBSeekCur {}\n", .{@intFromEnum(ibstream.IStreamSeekMode.kIBSeekCur)});
    try stdout.print("IBStream.kIBSeekEnd {}\n", .{@intFromEnum(ibstream.IStreamSeekMode.kIBSeekEnd)});
    try printTuid(stdout, "IBStream", ibstream.ibstream_iid);
    try printTuid(stdout, "ISizeableStream", ibstream.isizeable_stream_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime T: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(T), @alignOf(T) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
